File: reference/sdk/src/wallet.ts
import { ethers } from "ethers";
import { Context, ContractAddresses, Intent, TrustScore, NETWORKS } from "./types";
import { PolicyClient } from "./policy";
import { TrustClient } from "./trust";
import { IntentAttestationClient } from "./intent";

const WALLET_ABI = [
  "function openContext(bytes32 contextId, string calldata taskType) external",
  "function closeContext() external",
  "function executeSpend(address payable recipient, uint256 amount, string calldata category, bytes32 attestationId) external",
  "function getTrustScore() external view returns (uint256)",
  "function getActiveContext() external view returns (bytes32, string memory, uint256, bool)",
  "function updatePolicy(bytes32 newPolicyId) external",
  "function contextOpen() external view returns (bool)",
  "event SpendExecuted(address indexed recipient, uint256 amount, string category, bytes32 attestationId)",
  "event ContextOpened(bytes32 indexed contextId, string taskType, uint256 timestamp)",
  "event ContextClosed(bytes32 indexed contextId, uint256 timestamp)",
  "receive() external payable",
];

const WALLET_FACTORY_ABI = [
  "function createWallet() external returns (address)",
  "function getWallets(address owner) external view returns (address[])",
  "event WalletCreated(address indexed owner, address indexed walletAddress)",
];

const INTENT_ATTESTATION_ABI = [
  "function attest(bytes32 attestationId, string calldata action, string calldata reason, address recipient, uint256 amount) external",
  "function verify(bytes32 attestationId, address wallet) external view returns (bool)",
  "function getAttestation(bytes32 attestationId) external view returns (bytes32, address, string memory, string memory, address, uint256, uint256)",
  "event AttestationCreated(bytes32 indexed id, address indexed wallet, string action, string reason, address recipient, uint256 amount, uint256 timestamp)",
];

function resolveContracts(network: string): ContractAddresses {
  const net = NETWORKS[network];
  if (!net) throw new Error(`Unknown network: ${network}. Valid: ${Object.keys(NETWORKS).join(", ")}`);
  return net.contracts;
}

function getNextLevelAt(score: number): number {
  if (score < 100) return 100;
  if (score < 300) return 300;
  if (score < 600) return 600;
  if (score < 800) return 800;
  return 0; // autonomous — no next level
}

export class ARC402WalletClient {
  private walletContract: ethers.Contract;
  private walletAddress: string;
  private signer: ethers.Signer;
  private contracts: ContractAddresses;
  private intentContract: ethers.Contract;

  public policy: PolicyClient;
  public trust: TrustClient;
  public intent: IntentAttestationClient;

  constructor(
    walletAddress: string,
    signer: ethers.Signer,
    network: string = "base-sepolia"
  ) {
    this.walletAddress = walletAddress;
    this.signer = signer;
    this.contracts = resolveContracts(network);
    this.walletContract = new ethers.Contract(walletAddress, WALLET_ABI, signer);
    this.intentContract = new ethers.Contract(
      this.contracts.intentAttestation,
      INTENT_ATTESTATION_ABI,
      signer
    );
    this.policy = new PolicyClient(this.contracts.policyEngine, signer);
    this.trust = new TrustClient(this.contracts.trustRegistry, signer);
    this.intent = new IntentAttestationClient(
      this.contracts.intentAttestation,
      signer,
      walletAddress
    );
  }

  /** Deploy a new ARC-402 wallet via the WalletFactory and return a client for it. */
  static async deploy(
    signer: ethers.Signer,
    network: string = "base-sepolia"
  ): Promise<ARC402WalletClient> {
    const contracts = resolveContracts(network);
    if (!contracts.walletFactory) {
      throw new Error(`WalletFactory not deployed on ${network}`);
    }
    const factory = new ethers.Contract(contracts.walletFactory, WALLET_FACTORY_ABI, signer);
    const tx = await factory.createWallet();
    const receipt = await tx.wait();

    const iface = new ethers.Interface(WALLET_FACTORY_ABI);
    for (const log of receipt.logs) {
      try {
        const parsed = iface.parseLog(log);
        if (parsed && parsed.name === "WalletCreated") {
          return new ARC402WalletClient(parsed.args.walletAddress, signer, network);
        }
      } catch {}
    }
    throw new Error("Could not parse wallet address from WalletCreated event");
  }

  // ── Policy ──────────────────────────────────────────────────────────────────

  async setPolicy(categories: Record<string, bigint>): Promise<void> {
    await this.policy.set(this.walletAddress, categories);
  }

  async getPolicy() {
    return this.policy.get(this.walletAddress);
  }

  // ── Context ──────────────────────────────────────────────────────────────────

  async openContext(
    taskType: string
  ): Promise<{ contextId: string; close: () => Promise<void> }> {
    const contextId = ethers.keccak256(
      ethers.toUtf8Bytes(`${taskType}:${Date.now()}:${this.walletAddress}`)
    );
    const tx = await this.walletContract.openContext(contextId, taskType);
    await tx.wait();
    const self = this;
    return {
      contextId,
      close: () => self.closeContext(),
    };
  }

  async closeContext(): Promise<void> {
    const tx = await this.walletContract.closeContext();
    await tx.wait();
  }

  async getActiveContext(): Promise<Context | null> {
    const [contextId, taskType, openedAt, isOpen] =
      await this.walletContract.getActiveContext();
    if (!isOpen) return null;
    return {
      contextId,
      taskType,
      openedAt: Number(openedAt),
      isOpen,
    };
  }

  // ── Spend ────────────────────────────────────────────────────────────────────

  /**
   * Create an intent attestation, then execute the spend through the wallet contract.
   * Returns the attestationId (hex string).
   */
  async spend(
    recipient: string,
    amount: bigint,
    category: string,
    action: string,
    reason: string
  ): Promise<string> {
    const attestationId = ethers.keccak256(
      ethers.toUtf8Bytes(`${action}:${reason}:${recipient}:${amount}:${Date.now()}`)
    );

    const attestTx = await this.intentContract.attest(
      attestationId,
      action,
      reason,
      recipient,
      amount
    );
    await attestTx.wait();

    const tx = await this.walletContract.executeSpend(
      recipient,
      amount,
      category,
      attestationId
    );
    await tx.wait();
    return attestationId;
  }

  // ── Trust ────────────────────────────────────────────────────────────────────

  async getTrustScore(): Promise<TrustScore> {
    const score = await this.walletContract.getTrustScore();
    const numScore = Number(score);
    return {
      score: numScore,
      level: this.trust.getTrustLevel(numScore),
      nextLevelAt: getNextLevelAt(numScore),
    };
  }

  // ── Attestation history ───────────────────────────────────────────────────────

  async getAttestations(limit: number = 50): Promise<Intent[]> {
    const provider = this.signer.provider;
    if (!provider) return [];

    const iface = new ethers.Interface(INTENT_ATTESTATION_ABI);
    const eventFragment = iface.getEvent("AttestationCreated");
    if (!eventFragment) return [];

    const walletTopic = ethers.zeroPadValue(this.walletAddress.toLowerCase(), 32);
    const filter = {
      address: this.contracts.intentAttestation,
      topics: [
        eventFragment.topicHash,
        null,
        walletTopic,
      ],
      fromBlock: 0,
    };

    const logs = await provider.getLogs(filter);
    return logs.slice(-limit).map((log) => {
      const parsed = iface.parseLog(log)!;
      return {
        attestationId: parsed.args.id as string,
        action: parsed.args.action as string,
        reason: parsed.args.reason as string,
        recipient: parsed.args.recipient as string,
        amount: BigInt(parsed.args.amount),
        wallet: parsed.args.wallet as string,
        timestamp: Number(parsed.args.timestamp),
      };
    });
  }
}

// Aliases for backward compat
export class ARC402Wallet extends ARC402WalletClient {}

export class ContextBinding {
  private walletContract: ethers.Contract;

  constructor(walletAddress: string, signer: ethers.Signer) {
    this.walletContract = new ethers.Contract(walletAddress, WALLET_ABI, signer);
  }

  async open(taskType: string): Promise<string> {
    const contextId = ethers.keccak256(
      ethers.toUtf8Bytes(`${taskType}:${Date.now()}`)
    );
    const tx = await this.walletContract.openContext(contextId, taskType);
    await tx.wait();
    return contextId;
  }

  async close(): Promise<void> {
    const tx = await this.walletContract.closeContext();
    await tx.wait();
  }
}

File: python-sdk/arc402/wallet.py
"""ARC402Wallet — main entry point for the ARC-402 Python SDK."""

from __future__ import annotations

from eth_account import Account
from web3 import Web3

from .abis import ARC402Wallet_ABI, WalletFactory_ABI
from .context import ContextBinding
from .exceptions import ContextAlreadyOpen, ContextNotOpen, NetworkNotSupported, TransactionFailed
from .intent import IntentAttestation
from .policy import PolicyClient
from .settlement import MultiAgentSettlement
from .trust import TrustClient
from .types import AttestationRecord, NETWORKS, TrustScore


def _parse_amount(amount: str | int) -> int:
    if isinstance(amount, int):
        return amount
    if isinstance(amount, str):
        if "ether" in amount:
            return Web3.to_wei(amount.replace("ether", "").strip(), "ether")
        if "gwei" in amount:
            return Web3.to_wei(amount.replace("gwei", "").strip(), "gwei")
        return int(amount)
    raise TypeError(f"Cannot parse amount: {amount!r}")


class ARC402Wallet:
    def __init__(self, address: str, private_key: str, network: str = "base-sepolia"):
        if network not in NETWORKS:
            raise NetworkNotSupported(f"Network '{network}' is not supported. Choose from: {list(NETWORKS)}")

        net = NETWORKS[network]
        self._w3 = Web3(Web3.HTTPProvider(net["rpc_url"]))
        self._account = Account.from_key(private_key)
        self.address = Web3.to_checksum_address(address)
        self.network = network
        self._net_config = net

        self._wallet_contract = self._w3.eth.contract(address=self.address, abi=ARC402Wallet_ABI)
        self._policy = PolicyClient(self._w3, net["policy_engine"], self._account)
        self._trust = TrustClient(self._w3, net["trust_registry"], self._account)
        self._intent = IntentAttestation(self._w3, net["intent_attestation"], self._account)
        self._settlement = MultiAgentSettlement(self._w3, net["settlement_coordinator"], self._account)
        self._active_context_id: bytes | None = None

    @classmethod
    async def deploy(cls, private_key: str, network: str = "base-sepolia") -> "ARC402Wallet":
        if network not in NETWORKS:
            raise NetworkNotSupported(f"Network '{network}' is not supported.")
        net = NETWORKS[network]
        factory_address = net.get("wallet_factory")
        if not factory_address or factory_address == "0x0000000000000000000000000000000000000000":
            raise NetworkNotSupported(f"WalletFactory not deployed on {network}")

        w3 = Web3(Web3.HTTPProvider(net["rpc_url"]))
        account = Account.from_key(private_key)
        factory = w3.eth.contract(address=Web3.to_checksum_address(factory_address), abi=WalletFactory_ABI)
        tx = factory.functions.createWallet().build_transaction({
            "from": account.address,
            "nonce": w3.eth.get_transaction_count(account.address),
            "gas": 2_000_000,
            "gasPrice": w3.eth.gas_price,
            "chainId": w3.eth.chain_id,
        })
        signed = account.sign_transaction(tx)
        tx_hash = w3.eth.send_raw_transaction(signed.raw_transaction)
        receipt = w3.eth.wait_for_transaction_receipt(tx_hash)

        wallet_address = None
        for log in receipt.get("logs", []):
            if len(log.get("topics", [])) >= 3:
                wallet_address = "0x" + log["topics"][2].hex()[-40:]
                break
        if not wallet_address:
            raise TransactionFailed("Could not extract wallet address from factory receipt", tx_hash.hex())
        return cls(address=wallet_address, private_key=private_key, network=network)

    def context(self, task_type: str, task_id: str | None = None) -> ContextBinding:
        return ContextBinding(self, task_type=task_type, task_id=task_id)

    async def set_policy(self, config: dict[str, str | int]) -> str:
        return await self._policy.set_policy(self.address, config)

    async def spend(self, recipient: str, amount: str | int, category: str, reason: str) -> str:
        if self._active_context_id is None:
            raise ContextNotOpen("No active context. Use 'async with wallet.context(...)' before spending.")
        wei_amount = _parse_amount(amount)
        await self._policy.validate_spend(self.address, category, wei_amount, self._active_context_id)
        attestation_id = await self._intent.attest(action=f"spend:{category}", reason=reason, recipient=recipient, amount=wei_amount)
        tx = self._wallet_contract.functions.executeSpend(Web3.to_checksum_address(recipient), wei_amount, category, attestation_id).build_transaction(self._tx_params())
        receipt = await self._send(tx)
        return receipt["transactionHash"].hex()

    async def trust_score(self) -> TrustScore:
        return await self._trust.get_score(self.address)

    async def attestations(self, limit: int = 10) -> list[AttestationRecord]:
        return await self._intent.history(self.address, limit=limit)

    async def _open_context(self, context_id: bytes, task_type: str) -> None:
        if self._wallet_contract.functions.contextOpen().call():
            raise ContextAlreadyOpen("A context is already open on this wallet.")
        await self._send(self._wallet_contract.functions.openContext(context_id, task_type).build_transaction(self._tx_params()))
        self._active_context_id = context_id

    async def _close_context(self, success: bool = True) -> None:
        try:
            await self._send(self._wallet_contract.functions.closeContext().build_transaction(self._tx_params()))
        finally:
            self._active_context_id = None
            if success:
                await self._trust.record_success(self.address)
            else:
                await self._trust.record_anomaly(self.address)

    def _tx_params(self) -> dict:
        return {
            "from": self._account.address,
            "nonce": self._w3.eth.get_transaction_count(self._account.address),
            "gas": 500_000,
            "gasPrice": self._w3.eth.gas_price,
            "chainId": self._w3.eth.chain_id,
        }

    async def _send(self, tx: dict) -> dict:
        signed = self._account.sign_transaction(tx)
        tx_hash = self._w3.eth.send_raw_transaction(signed.raw_transaction)
        return self._w3.eth.wait_for_transaction_receipt(tx_hash)

    @property
    def policy(self) -> PolicyClient:
        return self._policy

    @property
    def trust(self) -> TrustClient:
        return self._trust

    @property
    def intent(self) -> IntentAttestation:
        return self._intent

    @property
    def settlement(self) -> MultiAgentSettlement:
        return self._settlement
